tests/integration/test_tls.py

```diff
@@ -0,0 +1,383 @@
+    elif isinstance(entity, Unit):
+        return (
+            entity.workload_status == status.name
+            and entity.workload_status_message == status.message
+        )
+    else:
+        raise ValueError(f"entity type {type(entity)} is not allowed")
+
+
+@pytest.mark.skip_if_deployed
+@pytest.mark.abort_on_fail
+async def test_build_and_deploy(
+    ops_test: OpsTest, kyuubi_charm, charm_versions, s3_bucket_and_creds
+):
+    """Test building and deploying the charm without relation with any other charm."""
+    image_version = METADATA["resources"]["kyuubi-image"]["upstream-source"]
+    resources = {"kyuubi-image": image_version}
+    logger.info(f"Image version: {image_version}")
+
+    # Deploy the charm and wait for waiting status
+    logger.info("Deploying kyuubi-k8s charm...")
+    await ops_test.model.deploy(
+        kyuubi_charm,
+        resources=resources,
+        application_name=APP_NAME,
+        num_units=1,
+        series="jammy",
+        trust=True,
+    )
+
+    logger.info("Waiting for kyuubi-k8s app to be idle...")
+    await ops_test.model.wait_for_idle(
+        apps=[APP_NAME],
+        timeout=1000,
+    )
+    logger.info(f"State of kyuubi-k8s app: {ops_test.model.applications[APP_NAME].status}")
+
+    logger.info("Setting configuration for kyuubi-k8s charm...")
+    namespace = ops_test.model.name
+    username = "kyuubi-spark-engine"
+    await ops_test.model.applications[APP_NAME].set_config(
+        {"namespace": namespace, "service-account": username}
+    )
+
+    logger.info("Waiting for kyuubi-k8s app to be idle...")
+    await ops_test.model.wait_for_idle(
+        apps=[APP_NAME],
+        status="blocked",
+        timeout=1000,
+    )
+
+    # Assert that the charm is in blocked state, waiting for Integration Hub relation
+    assert check_status(
+        ops_test.model.applications[APP_NAME], Status.MISSING_INTEGRATION_HUB.value
+    )
+
+    logger.info("Deploying s3-integrator charm...")
+    await ops_test.model.deploy(**charm_versions.s3.deploy_dict()),
+
+    logger.info("Waiting for s3-integrator app to be idle...")
+    await ops_test.model.wait_for_idle(
+        apps=[APP_NAME, charm_versions.s3.application_name], timeout=1000
+    )
+
+    # Receive S3 params from fixture
+    endpoint_url = s3_bucket_and_creds["endpoint"]
+    access_key = s3_bucket_and_creds["access_key"]
+    secret_key = s3_bucket_and_creds["secret_key"]
+    bucket_name = s3_bucket_and_creds["bucket"]
+
+    logger.info("Setting up s3 credentials in s3-integrator charm")
+    s3_integrator_unit = ops_test.model.applications[charm_versions.s3.application_name].units[0]
+    action = await s3_integrator_unit.run_action(
+        action_name="sync-s3-credentials", **{"access-key": access_key, "secret-key": secret_key}
+    )
+    await action.wait()
+
+    logger.info("Waiting for s3-integrator app to be idle and active...")
```

Comment:
We'll have this setup done by a helper function once we get the expose-external PR merged. :partying_face: